MINOR: Add extended cluster readiness check for Connect's embedded Kafka cluster

C0urante opened the pull request:
We're still seeing some flaky test failures for the `OffsetsApiIntegrationTest` suite, but in a much smaller subset of cases:
- testAlterSinkConnectorOffsetsDifferentKafkaClusterTargeted: [1](https://ge.apache.org/s/ydrj6vsi2t5mw/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.OffsetsApiIntegrationTest/testAlterSinkConnectorOffsetsDifferentKafkaClusterTargeted()?top-execution=1), [2](https://ge.apache.org/s/ndddgm7c6ma3c/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.OffsetsApiIntegrationTest/testAlterSinkConnectorOffsetsDifferentKafkaClusterTargeted()?top-execution=1)
- testResetSinkConnectorOffsetsDifferentKafkaClusterTargeted: [1](https://ge.apache.org/s/ydrj6vsi2t5mw/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.OffsetsApiIntegrationTest/testResetSinkConnectorOffsetsDifferentKafkaClusterTargeted()?top-execution=1), [2](https://ge.apache.org/s/ndddgm7c6ma3c/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.OffsetsApiIntegrationTest/testResetSinkConnectorOffsetsDifferentKafkaClusterTargeted()?top-execution=1)
- testGetSinkConnectorOffsetsDifferentKafkaClusterTargeted: [1](https://ge.apache.org/s/ydrj6vsi2t5mw/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.OffsetsApiIntegrationTest/testGetSinkConnectorOffsetsDifferentKafkaClusterTargeted()?top-execution=1)

After examining log files, it looks like this is a genuine case of timeouts being too low; in the three failures I examined, the sink connector's consumer group was never able to form or handle offset commits because the separate Kafka cluster it targeted didn't have a group coordinator and was still creating the internal offsets topic.

Instead of just increasing timeouts, I'd like to add an enhanced cluster readiness check for our `EmbeddedKafkaCluster` class that's automatically performed on startup. This way, not only do we give the test cases above more time to run (assuming that a significant amount of their runtime is currently taken up by bringing up the separate Kafka cluster), we also have better insight into whether future failures are caused by broker startup issues or something else.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
MINOR: Add extended cluster readiness check for Connect's embedded Kafka cluster

C0urante commented:
@gharris1727 If you have a moment could you take a look? Trying to continue on my war against flaky tests.

C0urante commented:
I also believe this would fix the recent flaky failure we saw for the `RestExtensionIntegrationTest::testRestExtensionApi` test case [here](https://ge.apache.org/s/ndddgm7c6ma3c/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.RestExtensionIntegrationTest/testRestExtensionApi()?top-execution=1). The test creates an embedded Connect cluster with a REST extension that tries to read the list of connectors on the cluster while it's being registered. The failure I linked for the test was caused because the request to list the connectors timed out, because the Connect worker was still starting, which in turn happened because the underlying Kafka cluster appeared to still be starting. This test case wasn't improved by our health check endpoint and the way that we use it to ensure that our embedded Connect clusters are healthy after they've started because REST extension registration takes place during cluster startup, before we start querying the health check endpoint.

C0urante committed:
Add readiness check to EmbeddedKafkaCluster::restartOnlyBrokers

C0urante commented:
I'm less certain of this, but it also might be a fix for the recent flaky we saw for the `ConnectWorkerIntegrationTest:: testRequestTimeouts` test case [here](https://ge.apache.org/s/ydrj6vsi2t5mw/tests/task/:connect:runtime:test/details/org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest/testRequestTimeouts()?top-execution=1). And, if not, it would again at least help us distinguish between broker errors and other ones, so I've added the readiness check to `EmbeddedKafkaCluster::restartOnlyBrokers`.

gharris1727 reviewed (approved):
Thanks @C0urante for the fix!

I think my primary hesitation with approving this is that it will extend the duration of tests  by simply doing more stuff. But if this can improve the reliability of our tests, I think that is worth it. We can always profile this later to improve our time-to-first-topic/time-to-first-consumer-group to claw back any slowdown this change introduces.

LGTM!

C0urante commented:
I'm optimistic that this won't have a serious impact on test runtime, since most of the operations added in this PR are already performed by Connect clusters anyways: forming a group (of workers, not necessarily consumers), creating internal topics, and writing to and reading from them.

This doesn't apply to cases where we spin up an embedded Kafka cluster separately from a Connect cluster, but based on our current tests, those are the cases that would benefit the most from this change. 

All that said, definitely open to refining this logic further in the future if it does turn out that we see a significant bump in test runtime with it.

C0urante committed:
Add key to cluster readiness record

Some tests alter the default topic cleanup policy from delete to compact, and compacted topics require record keys. So, we alter the cluster readiness check to write a record with both a key and a value.

OmniaGM commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -153,6 +154,49 @@ public void start() {
+            admin.deleteConsumerGroups(Collections.singleton(consumerGroupId)).all().get(30, TimeUnit.SECONDS);
+            admin.deleteTopics(Collections.singleton(topic)).all().get(30, TimeUnit.SECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new AssertionError("Failed to clean up cluster health check resource(s)", e);
```
Any reason why not using `Assertions.assertDoesNotThrow`? 
```
try (Admin admin = createAdminClient()) {
            Assertions.assertDoesNotThrow(() -> admin.deleteConsumerGroups(Collections.singleton(consumerGroupId)).all().get(30, TimeUnit.SECONDS), "Failed to clean up consumer groups during cluster health check resource(s)");
            Assertions.assertDoesNotThrow(() -> admin.deleteTopics(Collections.singleton(topic)).all().get(30, TimeUnit.SECONDS), "Failed to clean up topics during cluster health check resource(s)");
        }
```

OmniaGM replied:
Same btw can be done for `produce` as well `assertDoesNotThrow(() -> producer.send(msg).get());` we are doing something similar in tools

C0urante replied:
For the first comment, we already wrap things in a try-with-resources block in order to automatically close the consumer when the scope is exited and it keeps things simpler to add a corresponding catch clause.

I guess we could do this for produce but I'd rather not touch on that logic since it's not modified by this PR. If you or someone else wants to do a comprehensive refactor for our tests to adapt this style I'd be happy to take a look, though!
